src/app/api/auth/auth.ts

```diff
@@ -0,0 +1,54 @@
+// SPDX-FileCopyrightText: 2024 PNED G.I.E.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+import { decrypt } from '@/utils/encryption';
+import { jwtDecode } from 'jwt-decode';
+import { Account, getServerSession } from 'next-auth';
+import type { JWT } from 'next-auth/jwt';
+import { ExtendedSession } from './auth.types';
+import { authOptions } from './config';
+
+export async function getToken(tokenType: 'access_token' | 'id_token') {
+  const session = (await getServerSession(authOptions)) as ExtendedSession;
+  if (session) {
+    const tokenDecrypted = decrypt(session[tokenType]!);
+    return tokenDecrypted;
+  }
+  return null;
+}
+
+export function completeTokenWithAccountInfo(token: JWT, account: Account): JWT {
+  return {
+    ...token,
+    decoded: jwtDecode(account.access_token!) as string,
+    access_token: account.access_token as string,
+    id_token: account.id_token as string,
+    refresh_token: account.refresh_token as string,
+    expires_at: account.expires_at as number,
+  };
+}
+
+export async function refreshAccessToken(token: JWT) {
+  const response = await fetch(`${process.env.REFRESH_TOKEN_URL}`, {
+    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
+    body: new URLSearchParams({
+      client_id: `${process.env.KEYCLOAK_CLIENT_ID}`,
+      client_secret: `${process.env.KEYCLOAK_CLIENT_SECRET}`,
+      grant_type: 'refresh_token',
+      refresh_token: token.refresh_token as string,
+    }),
+    method: 'POST',
+    cache: 'no-cache',
```

Comment:
NextJS was automatically caching the results of this request and the refresh token was not retrieved properly each time

Reply:
Exactly, the new token was not refreshed properly because of some caching. With `fetch`, caching is done automatically by Next.js, so you have to specify explicitly to not cache the results

Reply:
I don't know if there is a way to disable caching globally by default in the app though

Reply:
why don't we use axios, didn't inderpal say the other day that you or emi (can't remember who) had some problems with fetch?

Reply:
Valid point. The authentication was implemented 3 months ago and we did not switch to Axios back then (if I remember correctly). But to be consistent yep we should move everything to Axios